### PR TITLE
services: migrate all oneshot systemd services to type=simple

### PR DESCRIPTION
**Pull request by goodwinos**
Systemd type=simple services (with Restart=no) can run for
a short time and then exit, much like a type=oneshot service
but without the need for all the KillMode=none baggage.

Systemd has a much more hands-off approach to simple services,
including not minding if forked subprocesses continue after a
oneshot script has exited.

This essentially changes all oneshot services from:
Type=oneshot
KillMode=none

To:
Type=simple
Restart=no

QA passes for the logctl group but this needs careful review and
a lot of soak time. As such these changes are somewhat experimental
at this stage.

Related: Red Hat BZ#1942844 - systemd-analyze complains about deprecated KillMode directive
Related: Fedora BZ#1897945 - pcp pmlogger services yield systemd killmode warning

### CODE REVIEW

**Comment by kmcdonell** [comment]
LGTM, but my views about systemd are well known, so perhaps I'm not the most objective reviewer.
I still think we need a root-to-branch audit of all of the systemd stuff in PCP by someone who's a real systemd expert.  I'm still seeing sporadic failure to restart the primary pmlogger in QA, and although this is outside the scope of the changes in this PR it does suggest the integration of PCP with systemd is still not quite right.
On the up side, getting rid of KillMode=none seems like a step in the right direction.

**Review by kmcdonell — APPROVED** [review]
See my comment added a moment ago.

**Comment by goodwinos** [comment]
Thanks Ken - at this stage I'm intending to merge this early for pcp-5.3.3. That would give us a whole release cycle of soak and the opportunity to make further changes to our systemd configs and the logctl scripts.

That said, nuking killmode is good initial benefit (resolves 2 BZs). And using type=simple for our logctl service scripts seems to be a very natural fit compared to oneshot. I've been testing this on a 10 logger mini-farm for the past three days and it's stable - and qa -g logctl is consistently passing. Maybe we should merge it??

**Comment by kmcdonell** [comment]
Your call ... I'd wait for early in the cycle for 5.3.3, but that's me being cautious.

File: src/pmfind/pmfind.service.in
Comment (keszybz) [inline src/pmfind/pmfind.service.in:10]: Restart=no is the default, so there isn't any need to specify this.

Type=simple means that the service will exit as soon as the main process has exited, i.e. the process started with ExecStart=.

I assume that you are trying to keep the original design where the main process exits but the children should remain running. This is not going to work like this, because systemd will kill those children when cleaning up, i.e. immediately after the main process has exited.

You have three options:
- change the design so that the main process waits for the children to exit before exiting itself
- change the design so that the main main process let's systemd think that the one of the children is the new main process, by setting Type=forking and either GuessMainPID=true or providing a file with the PID of one of the children. I don't know how many children there might be. This would require one of them to be guaranteed to run longer than the others.
- wait for https://github.com/systemd/systemd/issues/16805 to be implemented.

Please don't merge this as is.

(It doesn't matter here, since there are other issues, but `Type=simple` has been mostly obsoleted by `Type=exec`. Please use the latter whereever you'd use the former.)

Comment (gdamjan) [reply src/pmfind/pmfind.service.in:10]: another option, use something like `systemd-run` to run the children, so they'd be their own separate services.

Comment (goodwinos) [reply src/pmfind/pmfind.service.in:10]: Thanks for the comments - there is a bit of confusion here: not changing any of the PCP long running daemon services. They remain Type=notify. It's only the half hour pmlogger_check and nightly pmlogger_daily scripts (and similar for pmie and pmfind) services that are being changed. These services are started by a timer to perform regular house-keeping / log rolling etc and exit normally when completed, leaving no children. See new PR #1391 with improved commit messages. Not merging and closing this PR.

**Comment by goodwinos** [comment]
Closing, as described above.

**Pull request closed without merging**
